### ISSUE TITLE
W4fp8 AWQ

### ISSUE THREAD

**Issue by LugerW-A** [vllm-project/llm-compressor#1657]
W4fp8 achieves good quantization and acceleration effects. Can LLMCompressor implement this quantization method and adapt it to vLLM?

Ref:
https://nvidia.github.io/TensorRT-Model-Optimizer/guides/_choosing_quant_methods.html#:~:text=Ampere%20and%20later.-,INT4%2DFP8%20AWQ%20(W4A8),-High

**Comment by brian-dellabetta** [comment]
Hi @LugerW-A , my understanding from the AWQ work is that W4A8 won't work well for AWQ, because AWQ moves the variance into the activations so weights can be quantized. But maybe A8 is still high enough precision to make this ok? Seems like `modelopt` provides it without much issue, and I don't believe we have to do anything other than loosening [these validation checks](https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/modifiers/awq/base.py#L191-L204) to allow for 8-bit as well.

Feel free to try it out. I have some other priorities but can possibly take a look next week.

**Comment by LugerW-A** [comment]
@brian-dellabetta Thank you very much. My understanding is that Fp8 offers higher practical precision compared to Int8. W4Fp8 can effectively meet the production requirements for model accuracy, as well as prefill and decode inference performance.

**Comment by brian-dellabetta** [comment]
Would you have time to experiment and validate? I can set up a PR later this week, making sure it can run AWQ with W4A8 (int8 or fp8), and pass it back to you to try? I am unsure of the state of W4A8 support in vllm, but I believe it could at least be activated in `llm-compressor` fairly easily

**Comment by LugerW-A** [comment]
Okay, I will test the performance on both VLM and LLM

**Comment by LugerW-A** [comment]
> I am unsure of the state of W4A8 support in vllm

I hope the process of integrating this algorithm with vLLM isn't too difficult.
Maybe this pr can help https://github.com/vllm-project/vllm/pull/9855

**Comment by brian-dellabetta** [comment]
Hi @LugerW-A , had some time to dig in today. I've created a PR to allow for activation quantization:

* #1682

Can you check out branch `bdellabe/awq-w4a8` and try this out? A quick preliminary check with "meta-llama/Llama-3.2-3B-Instruct" seems promising 👀 :

```
Wikitext PPL (lower is better)
W4AFP8 AWQ+Quant --  13.0959
W4AFP8 Quant -- 13.2527
W4A16 AWQ+Quant --  13.1008
W4A16 Quant -- 13.2668
```

We have a `W4A8` preset quantization scheme, which uses integer quantization, but no `W4FP8` scheme. For that you can replace scheme with a `config_groups` field like so:
```python
from compressed_tensors.quantization import (
    QuantizationArgs,
    QuantizationScheme,
    QuantizationStrategy,
    QuantizationType,
)
recipe = [
    AWQModifier(
        ...
        config_groups={
            "group_0": QuantizationScheme(
                targets=["Linear"],
                weights=QuantizationArgs(
                    num_bits=4,
                    type=QuantizationType.INT,
                    group_size=128,
                    strategy=QuantizationStrategy.GROUP,
                    symmetric=True,
                    dynamic=False,
                ),
                input_activations=QuantizationArgs(
                    num_bits=8,
                    type=QuantizationType.FLOAT,
                    strategy=QuantizationStrategy.TOKEN,
                    symmetric=True,
                    dynamic=True,
                    observer=None,
                ),
            )
        },
    )
]
```

I unfortunately wasn't able to get the W4AFP4 model to load up in `vllm` though, it says not `compressed-tensors` compatible format was found. I'll ask around about the status of W4A8 support, I could very well be doing something wrong.



<details>
<summary>Full script I'm using:</summary>

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

from llmcompressor import oneshot, active_session
from llmcompressor.utils import dispatch_for_generation
from llmcompressor.modifiers.awq import AWQModifier, AWQMapping
from llmcompressor.modifiers.quantization import QuantizationModifier
from compressed_tensors.quantization import (
    QuantizationArgs,
    QuantizationScheme,
    QuantizationStrategy,
    QuantizationType,
)

# Select model and load it.
MODEL_ID = "meta-llama/Llama-3.2-3B-Instruct"

SAVE_DIR = MODEL_ID.split("/")[-1] + "-awq-asym"


# Configure the quantization algorithm to run.
recipe = [
    AWQModifier(
        ignore=[
            "lm_head",
            "re:.*mlp.gate$",
            "re:.*mlp.shared_expert_gate$",
            "re:visual.*",
        ],
        # scheme="W4A16",
        config_groups={
            "group_0": QuantizationScheme(
                targets=["Linear"],
                weights=QuantizationArgs(
                    num_bits=4,
                    type=QuantizationType.INT,
                    group_size=128,
                    strategy=QuantizationStrategy.GROUP,
                    symmetric=True,
                    dynamic=False,
                ),
                input_activations=QuantizationArgs(
                    num_bits=8,
                    type=QuantizationType.FLOAT,
                    strategy=QuantizationStrategy.TOKEN,
                    symmetric=True,
                    dynamic=True,
                    observer=None,
                ),
            )
        },
        targets=["Linear"],
    ),
]

# Select calibration dataset.
DATASET_ID = "mit-han-lab/pile-val-backup"
DATASET_SPLIT = "validation"

# Select number of samples. 256 samples is a good place to start.
# Increasing the number of samples can improve accuracy.
NUM_CALIBRATION_SAMPLES = 128  # 256
MAX_SEQUENCE_LENGTH = 512


def get_calib_dataset(tokenizer):
    from datasets import load_dataset

    ds = load_dataset(
        DATASET_ID,
        split=f"{DATASET_SPLIT}[:{NUM_CALIBRATION_SAMPLES*10}]",
    )

    def preprocess(example):
        return {"input_ids": tokenizer.encode(example["text"].strip())}

    ds = (
        ds.shuffle(seed=42)
        .map(preprocess, remove_columns=ds.column_names)
        .select(range(NUM_CALIBRATION_SAMPLES))
    )

    return ds


if __name__ == "__main__":
    model = AutoModelForCausalLM.from_pretrained(
        MODEL_ID, torch_dtype="auto", trust_remote_code=True
    )
    tokenizer = AutoTokenizer.from_pretrained(MODEL_ID, trust_remote_code=True)

    ###
    ### Apply algorithms.
    ###
    oneshot(
        model=model,
        dataset=get_calib_dataset(tokenizer),
        recipe=recipe,
        max_seq_length=MAX_SEQUENCE_LENGTH,
        num_calibration_samples=NUM_CALIBRATION_SAMPLES,
        log_dir=None,
        trust_remote_code_model=True,
    )

    # Confirm generations of the quantized model look sane.
    dispatch_for_generation(model)
    print("\n\n")
    print("========== SAMPLE GENERATION ==============")
    input_ids = tokenizer("Hello my name is", return_tensors="pt").input_ids.to("cuda")
    output = model.generate(input_ids, max_new_tokens=100)
    print(tokenizer.decode(output[0]))
    print("==========================================\n\n")

    # Save to disk compressed.
    model.save_pretrained(SAVE_DIR)
    tokenizer.save_pretrained(SAVE_DIR)

    ###
    #### Apply algorithms.
    ###

    ### LM EVAL

    active_session().reset()
    del model
    del tokenizer
    torch.cuda.empty_cache()

    import lm_eval
    from lm_eval.utils import make_table

    results = lm_eval.simple_evaluate(
        model="hf",
        model_args={
            "pretrained": SAVE_DIR,
            "add_bos_token": True,
            "dtype": "bfloat16",
        },
        tasks=["wikitext"],
        num_fewshot=5,
        batch_size=128,
    )
    print(make_table(results))
```

<br>
</details>

**Comment by brian-dellabetta** [comment]
Quick update on `vllm` W4A8 support, couple responses from the team:

>Towards support, w4a8 for fp isn't supported and I don't know of any work any time soon to support that. There was some external work contributed for w4a8 int, but that hasn't gone through significant performance testing and I believe was only hooked up for a specific input format that wasn't compressed tensors

---
>Hey Brian the PR you listed is just for a basic kernel support in Machete for the format, but we haven’t ever hooked it up end to end with a format. That is why “end2end integration” is a TODO
AFAIK the only w4a8 format we support at the moment is the QQQ format https://huggingface.co/HandH1998/QQQ-Llama-3-8b-g128
Generally we found that w4a8 has too much compute overhead to be competitive with w8a8 in throughput scenarios on hardware >=H100, so we haven’t invested in it much
There is some recent renewed interest in it specifically for MoE layers though

**Comment by LugerW-A** [comment]
Thanks! I'll try it.

**Comment by Bluedyson** [comment]
I tried using it, but I found an issue with the w4fp8 forward process. The weight is not actually computed in fp8 — it only converts from int4 to fp16. This may lead to incorrect accuracy results, as the actual impact of using fp8 on model accuracy is not properly reflected.


https://github.com/neuralmagic/compressed-tensors/blob/c6491592f9655a4197c2fc17335f6c0936b1ab74/src/compressed_tensors/quantization/lifecycle/forward.py#L375-L384
```python
        if scheme.input_activations is not None:
            # prehook should calibrate activations before forward call
            input_ = forward_quantize(module, input_, "input", scheme.input_activations)


        if scheme.weights is not None and not compressed:
            # calibrate and (fake) quantize weights when applicable
            unquantized_weight = self.weight.data.clone()
            self.weight.data = forward_quantize(
                module, self.weight, "weight", scheme.weights  # <- only weight int4 convert fp16, not have impact of fp8
            )
```

**Comment by brian-dellabetta** [comment]
Hi @Bluedyson , thanks for the message. Yeah this is a design choice and a bit of a grey area given it's beyond the original intentions of AWQ. We are seeking to minimize quantization loss with respect to the unquantized fp16 model, which I believe should stay in fp16. In order to compare the activations from modules where scales have been applied, they also need to remain in fp16 (or we could move to fp8 and upconvert to fp16 for the comparison). We could toggle this with a flag, but given how little changes going from fp16 to fp8 I don't expect it will change much. We'd want to do a more full-scale ablation study to confirm. I was hoping to hear from community users interested in this change before deciding and landing this PR

Have you had a chance to try it out? This is lower priority at the moment, but would definitely be curious to hear if it is working out for users.

**Comment by Bluedyson** [comment]
@brian-dellabetta Thank you for your attention. Over the past few days, I have updated the implementation of `compress-tensor` by introducing FP8 quantization error for weights. It has been observed that this approach achieves relatively good accuracy preservation in larger-scale models (32B). On our internal dataset, the KL divergence is 81 for W4A16 and 91 for W4A8, which still lags behind the 88 achieved by NVIDIA’s FP4A8-AWQ implemented in TensorRT-Model-Optimizer. However, it remains worthwhile on hardware that does not support NVFP4.  

Although this method did not yield satisfactory results on smaller models (0.6B), which was expected, I believe this feature is still necessary. Running W4A8 on graphics cards that do not support NVFP4—such as Hopper—can deliver significant performance improvements. Recently, I have also been working on adapting the kernel for SM89-based graphics cards. If VLLM could provide FP8 support for SM89, it would greatly enhance the performance of low-cost computing resources.

**Comment by Bluedyson** [comment]
@brian-dellabetta Thank you for your attention. Over the past few days, I have updated the implementation of `compress-tensor` by introducing FP8 quantization error for weights. The results show promising accuracy retention on larger-scale models (32B). On our internal dataset, the KL divergence is 81 for W4A16 and 91 for W4A8, which still lags behind the 88 achieved by NVIDIA’s FP4A8-AWQ implemented with TensorRT-Model-Optimizer. However, this approach remains worthwhile on hardware that does not support NVIDIA FP4, even though it did not yield satisfactory results on smaller models (0.6B), which was expected.  

I believe this feature is necessary, as running W4A8 on GPUs without NVIDIA FP4 support (such as the H20) can deliver higher performance. Specifically, in MoE layers, it significantly reduces memory bandwidth consumption compared to W8A8. Recently, I have also been working on adapting the kernel for SM89 GPUs. If VLLM could add FP8 support for SM89, it would greatly enhance the performance of cost-effective computing solutions.

**Comment by brian-dellabetta** [comment]
Hi @Bluedyson , thanks for all the detail! I need to update #1682  and get it merged in, but if you have any changes to `compressed-tensors` or `llm-compressor` that would be valuable to other users, please open a PR and I am happy to take a look. Did you just update the MSE loss in AWQ to compare against the fp8 output to the fp16 when finding best scales? The accuracy retention you're reporting on larger models is similar to what we've seen as well, smaller models (<1B) seem to collapse pretty quickly with stronger compression. 

> If VLLM could add FP8 support for SM89, it would greatly enhance the performance of cost-effective computing solutions.

I think this is unfortunately constrained at the hardware level? Or are you able to run NVIDIA's FP4A8 checkpoint on your system?

**Comment by Bluedyson** [comment]
@brian-dellabetta , thank you for your ongoing attention. Regarding compressed-tensors, I've updated the part concerning the quantization error of weights in fp8. I will organize these changes and submit a PR in the next couple of days. The point you mentioned was an oversight on my part—I had only adjusted the weights to fp8 quantization during inference, which already achieved what we considered a usable state, so I forget modify the MSE loss part. Over the past few days, I updated the MSE loss and found that w4fp8-AWQ has nearly reached the accuracy of nvfp4-fp8, which is very exciting. I will also organize this modification into a PR in the next couple of days.

Yes, nvfp4a8 indeed cannot run on the L20, but I noticed that TensorRT-LLM has implemented a kernel supporting weight-int4 and activation-fp8 on sm89, and it has achieved significant speedup results. The quantization algorithm implemented with TensorRT-Model-Optimizer did not perform better than llm-compressor, so I am trying to integrate this solution into vllm under the current framework rather than switching to TensorRT-LLM.

https://github.com/NVIDIA/TensorRT-LLM/blob/f4d3be4bbc625ddca055854af3b4682679560e35/tensorrt_llm/_torch/modules/linear.py#L1480-L1534
```python
class W4A8_AWQ_LinearMethod(LinearMethodBase):
    def apply(self, module: Linear, input: torch.Tensor,
              bias: Optional[torch.Tensor]):
        """
        modelopt flow for w4a8_awq:
         1. multiply pre_quant_scale to input
         2. quantize input to fp8 using input_scale
         3. unpack_weights and multiply by weight_scales (int4 -> fp16)
         4. divied by weight_scale_2 (fp16 -> fp8 to allow gemm in fp8).
         5. apply gemm in fp8.
         6. rescale using alpha which is input_scale * weight_scale_2
        """
```

**Comment by brian-dellabetta** [comment]
Thanks @Bluedyson for the detail! Will share this with the team next week, and keep an eye out for the PR.

> w4fp8-AWQ has nearly reached the accuracy of nvfp4-fp8, which is very exciting

Awesome! we may be able to squeeze it even further, because NVFP4 is likely using a smaller group size.
